### PR TITLE
sync subtree

### PR DESCRIPTION
**Pull request by calebcartwright**
@ytmimi I'll be intermittently away from my machine today so if you're around it'd be helpful if you could keep an eye on status and ping me/merge as necessary

(though remember that for sync PRs we _never_ squash nor rebase the merge commits)

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -81,7 +81,11 @@
 - Support for formatting let-else statements [#5690]
 - New config option, `single_line_let_else_max_width`, that allows users to configure the maximum length of single line `let-else` statements. `let-else` statements that otherwise meet the requirements to be formatted on a single line will have their divergent`else` block formatted over multiple lines if they exceed this length [#5684]
 
+<<<<<<< HEAD
+[#5690]: (https://github.com/rust-lang/rustfmt/pulls/5690)
+=======
 [#5690]: https://github.com/rust-lang/rustfmt/pull/5690
+>>>>>>> upstream/master
 [#5684]: https://github.com/rust-lang/rustfmt/issues/5684
```
Comment (calebcartwright) [inline CHANGELOG.md:89]: Ugh :vomiting_face: 

I clearly got tired of merge conflicts. Mercifully this is just noise in the changelog, so i'll clean this up later when I add the additional entries

**Comment by calebcartwright** [comment]
It seems that non-`&&` operators fail during parsing now as opposed to ast validation stage, so we'll need to update the let chains tests accordingly

File: tests/target/let_chains.rs [hunk 2]
```diff
@@ -13,11 +13,12 @@ fn main() {
...
-        || aaaaaaaaa
-            && let Some(x) = xxxxxxxxxxxx
-            && aaaaaaa
-            && let None = aaaaaaaaaa
+    if aaaaaaaaaaaaaaaaaaaaa
+        && aaaaaaaaaaaaaaa
+        && aaaaaaaaa
+        && let Some(x) = xxxxxxxxxxxx
+        && aaaaaaa
+        && let None = aaaaaaaaaa
```
Comment (calebcartwright) [inline tests/target/let_chains.rs:21]: I do have to step away now for a while, but wanted to flag this for your attention. Given that all the operators are now the same this does appear to be the correct formatting to my mind, do you agree?

**Review by ytmimi — APPROVED** [review]

File: tests/target/let_chains.rs (hunk 2 above)
Comment (ytmimi) [reply tests/target/let_chains.rs:21]: Yeah, this is the formatting that I would have expected

File: CHANGELOG.md (hunk 1 above)
Comment (ytmimi) [reply CHANGELOG.md:89]: 🤣 no worries. I know you mentioned that there were a lot of conflicts. No worries on this!

**Comment by ytmimi** [comment]
Just for good measure I'm running the [Diff Check Job](https://github.com/rust-lang/rustfmt/actions/runs/6605606043)

**Comment by calebcartwright** [comment]
> Just for good measure I'm running the [Diff Check Job](https://github.com/rust-lang/rustfmt/actions/runs/6605606043)

Good, I'd been meaning to suggest we do so just to get a feel for the let chain results. Obviously we should be expecting some level of diff in just about every repo that has let chains in the code

**Comment by ytmimi** [comment]
I actually don't think there will be a change since `let-chain` support has already been merged, and we compare results between running the latest rustfmt from source and this PRs version of rustfmt.

**Comment by calebcartwright** [comment]
I'm going to go ahead and merge this then. If you have bandwidth now/within the next couple hours to do the version bump and changelog that'd also be a big help :pray: (a single commit that adds the new entries to the changelog + another pass to fix some typos, and then does a patch version bump in the cargo.toml + lockfile files would do it)

**Comment by ytmimi** [comment]
hmmm I guess 🤔 I spoke too soon: https://github.com/rust-lang/rustfmt/actions/runs/6605606043/job/17940945214

**Comment by ytmimi** [comment]
Going to look into this

**Comment by ytmimi** [comment]
@calebcartwright Okay, I think I know why the diff-check job failed, and I'm happy to say that it was a false positive. When I ran things locally, everything checked out 🙏🏼

You can inspect the zip if you'd like.

[diff-check.zip](https://github.com/rust-lang/rustfmt/files/13064499/diff-check.zip)


Here's the explanation for what went wrong.

When we're doing the subtree sync we're in an interesting scenario where rustfmt at `origin/master` and rustfmt from the feature branch (in this case `subtree-sync-2023-10-22`) need different `LD_LIBRARY_PATH`s to be set since the subtree sync modifies the `rust-toolchain` file.

I think the reason we see diffs in the output is because the binary we built from `subtree-sync-2023-10-22` failed to run at all, and we silence those issues with `set +e`:
https://github.com/rust-lang/rustfmt/blob/0bb2acf50fad389285d22b2706d592b9668e65d5/ci/check_diff.sh#L149-L152

I couldn't re-run the CI job directly since you merged and deleted the `subtree-sync-2023-10-22` branch, so I needed to work around that when running the script locally. Here are the steps you could take to reproduce the successful `diff-check` run:

1) created a [`subtree-sync-2023-10-22`](https://github.com/ytmimi/rustfmt/tree/subtree-sync-2023-10-22) branch in my fork of the repo where the HEAD is in line with `origin/master` (0bb2acf)
2) checked out [ytmimi:diff-check-export](https://github.com/ytmimi/rustfmt/tree/diff-check-export)
3) applied this patch to modify how we set `LD_LIBRARY_PATH` and hard code 547577fa5d309d90292ca3a58fef1bf0d9325cc0 as the master rustfmt:
    ```diff
    diff --git a/ci/check_diff.sh b/ci/check_diff.sh
    index b6ce62c2..c5ea2792 100755
    --- a/ci/check_diff.sh
    +++ b/ci/check_diff.sh
    @@ -2,9 +2,6 @@
 
     set -e
 
    -# https://github.com/rust-lang/rustfmt/issues/5675
    -export LD_LIBRARY_PATH=$(rustc --print sysroot)/lib:$LD_LIBRARY_PATH
    -
    function print_usage() {
         echo "usage check_diff REMOTE_REPO FEATURE_BRANCH [COMMIT_HASH] [OPTIONAL_RUSTFMT_CONFIGS]"
    }
    @@ -127,6 +124,13 @@ function compile_rustfmt() {
         echo -e "\ncompiling with $CARGO_VERSON\n"
 
         echo "Building rustfmt from src"
    +    # rustfmt commit before the latest subtree sync (https://github.com/rust-lang/rustfmt/pull/5944)
    +    echo "checking out commit hash 547577fa5d309d90292ca3a58fef1bf0d9325cc0"
    +    git fetch origin 547577fa5d309d90292ca3a58fef1bf0d9325cc0
    +    git checkout --detach 547577fa5d309d90292ca3a58fef1bf0d9325cc0
    +    # Add the LD_LIBRARY_PATH for master rustfmt
    +    export LD_LIBRARY_PATH=$(rustc --print sysroot)/lib:$LD_LIBRARY_PATH
    +    echo "LD_LIBRARY_PATH: $LD_LIBRARY_PATH"
         cargo build -q --release --bin rustfmt && cp target/release/rustfmt $1/rustfmt
 
         if [ -z "$OPTIONAL_COMMIT_HASH" ] || [ "$FEATURE_BRANCH" = "$OPTIONAL_COMMIT_HASH" ]; then
    @@ -136,6 +140,11 @@ function compile_rustfmt() {
         fi
 
         echo "Building feature rustfmt from src"
    +    # Add the `LD_LIBRARY_PATH` for feature rustfmt. In most cases the `LD_LIBRARY_PATH` should be the same,
    +    # for both binaries, however during subtree syncs we bump the `channel` in the rust-toolchain file, and
    +    # therefore the feature branch needs to use a different sysroot.
    +    export LD_LIBRARY_PATH=$(rustc --print sysroot)/lib:$LD_LIBRARY_PATH
    +    echo "LD_LIBRARY_PATH: $LD_LIBRARY_PATH"
         cargo build -q --release --bin rustfmt && cp target/release/rustfmt $1/feature_rustfmt
 
         RUSFMT_BIN=$1/rustfmt
    ```
4) run the script:
    ```
    ./ci/check_diff.sh https://github.com/ytmimi/rustfmt.git subtree-sync-2023-10-22 subtree-sync-2023-10-22
    ```

**Comment by calebcartwright** [comment]
Would it help if I restored the branch?

**Comment by ytmimi** [comment]
I don't think restoring the branch would have made much of a difference anyway.

The issue was that only the master rustfmt binary ran successfully. When it ran it produced diffs, but the feature branch binary failed to run because of the runtime dependency on a different sysroot so no diffs were produced, and the diff-check script saw that discrepancy and marked the job as a failure.

I'll open up a PR with a similar patch to avoid these issues for future sub tree syncs moving forward.

**Pull request merged**